// Floating Temple
// Copyright 2015 Derek S. Snyder
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import "engine/proto/uuid.proto";

package floating_temple.engine;

message EmptyValue {
}

message ValueProto {
  enum Type {
    UNKNOWN = 0;

    EMPTY = 1;
    DOUBLE = 2;
    FLOAT = 3;
    INT64 = 4;
    UINT64 = 5;
    BOOL = 6;
    STRING = 7;
    BYTES = 8;

    OBJECT_ID = 9;
  }

  required int64 local_type = 1;

  // Exactly one of the following fields must be set.
  optional floating_temple.engine.EmptyValue empty_value = 2;
  optional double double_value = 3;
  optional float float_value = 4;
  optional sint64 int64_value = 5;
  optional uint64 uint64_value = 6;
  optional bool bool_value = 7;
  optional string string_value = 8;
  optional bytes bytes_value = 9;
  optional floating_temple.engine.Uuid object_id = 10;
}
